apollo-router/src/services/subgraph_service.rs

```diff
@@ -602,100 +604,205 @@ async fn call_http(
-                {
+    // The graphql spec is lax about what strategy to use for processing responses: https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#processing-the-response
+    //
+    // "If the response uses a non-200 status code and the media type of the response payload is application/json
+    // then the client MUST NOT rely on the body to be a well-formed GraphQL response since the source of the response
+    // may not be the server but instead some intermediary such as API gateways, proxies, firewalls, etc."
+    //
+    // The TLDR of this is that it's really asking us to do the best we can with whatever information we have with some modifications depending on content type.
+    // Our goal is to give the user the most relevant information possible in the response errors
+    //
+    // Rules:
+    // 1. If the content type of the response is not `application/json` or `application/graphql-response+json` then we won't try to parse.
+    // 2. If an HTTP status is not 2xx it will always be attached as a graphql error.
+    // 3. If the response type is `application/json` and status is not 2xx and the body the entire body will be output if the response is not valid graphql.
 
-                    return if !parts.status.is_success() {
+    let display_headers = context.contains_key(LOGGING_DISPLAY_HEADERS);
+    let display_body = context.contains_key(LOGGING_DISPLAY_BODY);
 
-                        Err(BoxError::from(FetchError::SubrequestHttpError {
-                            service: service_name.clone(),
-                            status_code: Some(parts.status.as_u16()),
-                            reason: format!(
-                                "{}: {}",
-                                parts.status.as_str(),
-                                parts.status.canonical_reason().unwrap_or("Unknown")
-                            ),
-                        }))
-                    } else {
-                        Err(BoxError::from(FetchError::SubrequestHttpError {
-                            status_code: Some(parts.status.as_u16()),
-                            service: service_name.clone(),
-                            reason: format!("subgraph didn't return JSON (expected content-type: {} or content-type: {GRAPHQL_JSON_RESPONSE_HEADER_VALUE}; found content-type: {content_type:?})", APPLICATION_JSON.essence_str()),
-                        }))
-                    };
+    // Print out the debug for the request
+    if display_headers {
+        tracing::info!(http.request.headers = ?request.headers(), apollo.subgraph.name = %service_name, "Request headers to subgraph {service_name:?}");
+    }
+    if display_body {
+        tracing::info!(http.request.body = ?request.body(), apollo.subgraph.name = %service_name, "Request body to subgraph {service_name:?}");
+    }
+
+    // Perform the actual fetch. I this fails then
```

Comment:
if this fails then ? ^^

Reply:
[77ccb05](https://github.com/apollographql/router/pull/3328/commits/77ccb0577ba50ef3fe1b45d705fa12db7a58f9c8)